File: 0980-Unique-Paths-III/0980-Unique-Paths-III.py
class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        start_row, start_col, end_row, end_col = 0, 0, 0, 0
        empty_squares = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == 1:
                    start_row = i
                    start_col = j
                elif grid[i][j] == 2:
                    end_row = i
                    end_col = j
                elif grid[i][j] == 0:
                    empty_squares += 1
        return self.dfs(grid, start_row, start_col, end_row, end_col, empty_squares)
    
    def dfs(self, grid: List[List[int]], row: int, col: int, end_row: int, end_col: int, empty_squares: int) -> int:
        if row < 0 or row >= len(grid) or col < 0 or col >= len(grid[0]) or grid[row][col] == -1:
            return 0
        if grid[row][col] == 2:
            return 1 if empty_squares == -1 else 0
        grid[row][col] = -1
        empty_squares -= 1
        paths = self.dfs(grid, row + 1, col, end_row, end_col, empty_squares) \
            + self.dfs(grid, row - 1, col, end_row, end_col, empty_squares) \
            + self.dfs(grid, row, col + 1, end_row, end_col, empty_squares) \
            + self.dfs(grid, row, col - 1, end_row, end_col, empty_squares)
        grid[row][col] = 0
        empty_squares += 1
        return paths

File: 0980-Unique-Paths-III/0980-Unique-Paths-III.cpp
class Solution {
public:
    int uniquePathsIII(vector<vector<int>>& grid) {
        int startRow, startCol, endRow, endCol;
        int emptySquares = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[i].size(); j++) {
                if (grid[i][j] == 1) {
                    startRow = i;
                    startCol = j;
                } else if (grid[i][j] == 2) {
                    endRow = i;
                    endCol = j;
                } else if (grid[i][j] == 0) {
                    emptySquares++;
                }
            }
        }
        return dfs(grid, startRow, startCol, endRow, endCol, emptySquares);
    }
private:
    int dfs(vector<vector<int>>& grid, int row, int col, int endRow, int endCol, int emptySquares) {
        if (row < 0 || row >= grid.size() || col < 0 || col >= grid[0].size() || grid[row][col] == -1) {
            return 0;
        }
        if (grid[row][col] == 2) {
            return emptySquares == -1 ? 1 : 0;
            }
        grid[row][col] = -1;
        emptySquares--;
        int paths =dfs(grid, row + 1, col, endRow, endCol, emptySquares)\
        + dfs(grid, row - 1, col, endRow, endCol, emptySquares)\
        + dfs(grid, row, col + 1, endRow, endCol, emptySquares)\
        + dfs(grid, row, col - 1, endRow, endCol, emptySquares);
        grid[row][col] = 0;
        emptySquares++;
        return paths;
    }
};
